File: arduinobot_ws/src/arduinobot_test/scripts/simple_action_server.py
#! /usr/bin/env python3
import rospy
import actionlib
from arduinobot_test.msg import FibonacciAction, FibonacciResult, FibonacciFeedback


class FibonacciActionServer(object):
    # create messages that are used to publish feedback/result
    feedback_ = FibonacciFeedback()
    result_ = FibonacciResult()

    def __init__(self, name):
        # Constructor
        # function that inizialize the FibonacciAction class and creates 
        # a Simple Action Server from the library actionlib
        self.action_name_ = name
        self.as_ = actionlib.SimpleActionServer(self.action_name_, FibonacciAction, execute_cb=self.execute_cb, auto_start = False)
        self.as_.start()
        rospy.loginfo('Simple Action Server Started')
      

    def execute_cb(self, goal):
        rospy.loginfo('Goal Received : %s' % goal.order)

        # Function that is called each time the action server receives a new goal
        # It starts displaying a fibonacci series from 1 to the goal target
        r = rospy.Rate(1)
        success = True
        
        # Sends an action feedback 
        self.feedback_.sequence = []
        self.feedback_.sequence.append(1)
        self.feedback_.sequence.append(1)
        

        # start executing the action
        for i in range(1, (goal.order - 1)):
            # check that preempt has not been requested by the client
            if self.as_.is_preempt_requested():
                rospy.loginfo('%s: Preempted' % self.action_name_)
                self.as_.set_preempted()
                success = False
                break
            self.feedback_.sequence.append(self.feedback_.sequence[i] + self.feedback_.sequence[i-1])
            # publish the feedback
            self.as_.publishfeedback_(self.feedback_)
            # this step is not necessary, the sequence is computed at 1 Hz for demonstration purposes
            r.sleep()
          
        # If no cancelation requests are received, return a suceeded result 
        if success:
            self.result_.sequence = self.feedback_.sequence
            rospy.loginfo('%s: Succeeded' % self.action_name_)
            self.as_.set_succeeded(self.result_)


if __name__ == '__main__':

    # Inizialize a ROS node called fibonacci
    rospy.init_node('fibonacci')

    # create an instance of the class FibonacciAction which will inizialize
    # a Simple Action Server from the actionlib library called finonacci
    server = FibonacciActionServer(rospy.get_name())

    # keeps the node up and running
    rospy.spin()

File: arduinobot_ws/src/arduinobot_test/src/simple_action_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <arduinobot_test/FibonacciAction.h>


class FibonacciActionServer
{
protected:

  ros::NodeHandle nh_;
  actionlib::SimpleActionServer<arduinobot_test::FibonacciAction> as_; 
  std::string action_name_;
  // create messages that are used to published feedback/result
  arduinobot_test::FibonacciFeedback feedback_;
  arduinobot_test::FibonacciResult result_;

public:

  FibonacciActionServer(std::string name) :
    as_(nh_, name, boost::bind(&FibonacciActionServer::executeCB, this, _1), false),
    action_name_(name)
  {
    as_.start();
    ROS_INFO("Simple Action Server Started");
  }

  void executeCB(const arduinobot_test::FibonacciGoalConstPtr &goal)
  {
    ROS_INFO("Goal Received %i", goal->order);

    // helper variables
    ros::Rate r(1);
    bool success = true;

    // push_back the seeds for the fibonacci sequence
    feedback_.sequence.clear();
    feedback_.sequence.push_back(1);
    feedback_.sequence.push_back(1);

    // start executing the action
    for(int i=1; i<=(goal->order-1); i++)
    {
      // check that preempt has not been requested by the client
      if (as_.isPreemptRequested() || !ros::ok())
      {
        ROS_INFO("%s: Preempted", action_name_.c_str());
        // set the action state to preempted
        as_.setPreempted();
        success = false;
        break;
      }
      feedback_.sequence.push_back(feedback_.sequence[i] + feedback_.sequence[i-1]);
      // publish the feedback
      as_.publishFeedback(feedback_);
      // this sleep is not necessary, the sequence is computed at 1 Hz for demonstration purposes
      r.sleep();
    }

    if(success)
    {
      result_.sequence = feedback_.sequence;
      ROS_INFO("%s: Succeeded", action_name_.c_str());
      // set the action state to succeeded
      as_.setSucceeded(result_);
    }
  }


};


int main(int argc, char** argv)
{
  ros::init(argc, argv, "fibonacci");

  FibonacciActionServer fibonacci("fibonacci");
  ros::spin();

  return 0;
}